File: api/internal/model/service_history.go
package model

import "time"

type HistoryEntryType string

const (
	HistoryTypeService HistoryEntryType = "service"
	HistoryTypeRepair  HistoryEntryType = "repair"
)

type ServiceHistory struct {
	ID                string           `json:"id" bson:"_id,omitempty"`
	VehicleID         string           `json:"vehicle_id" bson:"vehicle_id"`
	Type              HistoryEntryType `json:"type" bson:"type"`
	MaintenanceItemID *string          `json:"maintenance_item_id" bson:"maintenance_item_id"`
	ItemName          string           `json:"item_name" bson:"item_name"`
	ServiceUsage      *float64         `json:"service_usage" bson:"service_usage"`
	ServiceDate       time.Time        `json:"service_date" bson:"service_date"`
	Cost              *float64         `json:"cost" bson:"cost"`
	Provider          *string          `json:"provider" bson:"provider"`
	LoggedBy          string           `json:"logged_by" bson:"logged_by"`
	CreatedAt         time.Time        `json:"created_at" bson:"created_at"`
}

type LogServiceRequest struct {
	MaintenanceItemID string   `json:"maintenance_item_id"`
	ServiceUsage      float64  `json:"service_usage"`
	ServiceDate       string   `json:"service_date"` // ISO 8601
	Cost              *float64 `json:"cost,omitempty"`
	Provider          *string  `json:"provider,omitempty"`
}

type LogRepairRequest struct {
	Description string   `json:"description"`
	ServiceDate string   `json:"service_date"` // ISO 8601
	Cost        *float64 `json:"cost,omitempty"`
	Provider    *string  `json:"provider,omitempty"`
}

type UpdateServiceHistoryRequest struct {
	ServiceUsage *float64 `json:"service_usage,omitempty"`
	ServiceDate  *string  `json:"service_date,omitempty"`
	Cost         *float64 `json:"cost,omitempty"`
	Provider     *string  `json:"provider,omitempty"`
	Description  *string  `json:"description,omitempty"` // For repairs only
}

File: mobile/lib/models/service_history.dart
enum HistoryEntryType { service, repair }

class ServiceHistory {
  final String id;
  final String vehicleId;
  final HistoryEntryType type;
  final String? maintenanceItemId;
  final String itemName;
  final double? serviceUsage;
  final DateTime serviceDate;
  final double? cost;
  final String? provider;
  final String loggedBy;
  final DateTime createdAt;

  ServiceHistory({
    required this.id,
    required this.vehicleId,
    required this.type,
    this.maintenanceItemId,
    required this.itemName,
    this.serviceUsage,
    required this.serviceDate,
    this.cost,
    this.provider,
    required this.loggedBy,
    required this.createdAt,
  });

  Map<String, dynamic> toJson() => {
        'id': id,
        'vehicle_id': vehicleId,
        'type': type == HistoryEntryType.repair ? 'repair' : 'service',
        if (maintenanceItemId != null) 'maintenance_item_id': maintenanceItemId,
        'item_name': itemName,
        if (serviceUsage != null) 'service_usage': serviceUsage,
        'service_date': serviceDate.toIso8601String(),
        if (cost != null) 'cost': cost,
        if (provider != null) 'provider': provider,
        'logged_by': loggedBy,
        'created_at': createdAt.toIso8601String(),
      };

  factory ServiceHistory.fromJson(Map<String, dynamic> json) => ServiceHistory(
        id: json['id'] as String,
        vehicleId: json['vehicle_id'] as String,
        type: json['type'] == 'repair'
            ? HistoryEntryType.repair
            : HistoryEntryType.service,
        maintenanceItemId: json['maintenance_item_id'] as String?,
        itemName: json['item_name'] as String,
        serviceUsage: (json['service_usage'] as num?)?.toDouble(),
        serviceDate: DateTime.parse(json['service_date'] as String),
        cost: (json['cost'] as num?)?.toDouble(),
        provider: json['provider'] as String?,
        loggedBy: json['logged_by'] as String,
        createdAt: DateTime.parse(json['created_at'] as String),
      );
}
